### ISSUE TITLE
Add more hinting modes

### ISSUE THREAD

**Issue by The-Compiler** [The-Compiler/qutebrowser#521]
From Vimperator:

```
this._hintModes = {
    ";": Mode("Focus hint",                         function (elem) buffer.focusElement(elem),                             extended),
    "?": Mode("Show information for hint",          function (elem) buffer.showElementInfo(elem),                          extended),
    s: Mode("Save link",                            function (elem) buffer.saveLink(elem, true)),
    S: Mode("Save object",                          function (elem) buffer.saveLink(elem, true),                           images),
    a: Mode("Save link with prompt",                function (elem) buffer.saveLink(elem, false)),
    A: Mode("Save object with prompt",              function (elem) buffer.saveLink(elem, false),                          images),
    f: Mode("Focus frame",                          function (elem) Buffer.focusedWindow = elem.ownerDocument.defaultView, function () util.makeXPath(["body"])),
    [...]
    v: Mode("View hint source",                     function (elem, loc) buffer.viewSource(loc, false),                    extended),
    V: Mode("View hint source in external editor",  function (elem, loc) buffer.viewSource(loc, true),                     extended),
    y: Mode("Yank hint location",                   function (elem, loc) util.copyToClipboard(loc, true)),
    Y: Mode("Yank hint description",                function (elem) util.copyToClipboard(elem.textContent || "", true),    extended),
    n: Mode("Yank hint anchor URL",                 function (elem) {
        let uri = elem.ownerDocument.documentURIObject.clone();
        uri.ref = elem.id || elem.name;
        util.copyToClipboard(uri.spec, true);
    }, anchors),
    c: Mode("Open context menu",                    function (elem) buffer.openContextMenu(elem),                          extended),
    i: Mode("Show media object",                    function (elem) liberator.open(elem.src),                              images),
    I: Mode("Show media object in a new tab",       function (elem) liberator.open(elem.src, liberator.NEW_TAB),           images),
    x: Mode("Show hint's title or alt text",        function (elem) liberator.echo(elem.title ? "title: " + elem.title : "alt: " + elem.alt), function() "//*[@title or @alt]")
}
```

Also see #446.

**Comment by The-Compiler** [comment]
We also could have "extended" modes which check all elements for javascript click handlers:

```
18:57 <Unison> qutebrowser is 99% perfect in this regard, but it wont help you follow a div with a click listener
18:59 <The-Compiler> Unison: do you have an example? I wonder if I can fix that...
19:07 <Unison> The-Compiler: <div>a</div> <script>$("div").click(function() { alert("example") });</script>
19:08 <Unison> ^ Hinting will ignore that div
19:13 <The-Compiler> Unison: could one detect such a click handler using js? I don't know js very well :)
19:15 <The-Compiler> Unison: and couldn't such a handler be on any element? Or is it typically on div-elements?
19:15 <Unison> The-Compiler: Yep it can be done. I want to add that at some point
19:16 <Unison> It could be on any arbitrary element
19:18 <iggy> js event handlers (onclick, etc.) can be on any element represented in the dom
19:19 <The-Compiler> Unison: cool! I'll open an issue so I don't forget. Also for the checkbox thing :)
19:19 <The-Compiler> iggy: Okay... I'd guess checking every element will be a performance nightmare though :-/
19:20 <The-Compiler> Unless you can select such elements with a css selector which I doubt
19:20 <iggy> heh, yeah, probably
19:25 <Unison> The-Compiler: You can't. I'd probably make a new hinting mode (i.e. 'deep hinting') to find those
19:30 <Unison> just for slower the machine's sake
19:30 <Unison> My plan was ;;h ;;y ;;r etc..
```

**Comment by The-Compiler** [comment]
A rapid hinting mode for yank also would be nice, which yanks the URLs newline-separated.

**Comment by The-Compiler** [comment]
Another nice thing would be a mode for the element the mouse is over currently.

**Comment by The-Compiler** [comment]
Another one: Yanking the `src` attribute of an image, like mentioned in #1206.

**Comment by AtzeDeVries** [comment]
also would like to have a mouse over (as mentioned on sept 6) hinting. It would be nice to have this in combination with 'normal' hinting.

For example, having a site with dropdown menu's where click on the menu and dropdown does something. In normal hinting mode a mouse over hint could be in a different color.

**Comment by The-Compiler** [comment]
@AtzeDeVries FWIW, you can already do mouseover hinting using `;h`.

**Comment by The-Compiler** [comment]
A mode which finds links hidden in text (without `<a>`) would be nice as well, but probably quite hard to implement.

**Comment by The-Compiler** [comment]
Apparently pentadactyl also has a pseudo-frames mode (`F`) which hints scrollable elements like the code block on https://developer.nvidia.com/accelerated-computing-training

![bq4](https://cloud.githubusercontent.com/assets/625793/17478207/68cf05e0-5d6a-11e6-9c48-bfbd780e8d87.png)

**Comment by The-Compiler** [comment]
Embedded youtube videos also sometimes don't have hints. Turns out there are probably four different ways of doing so:

- `<iframe>` (gets hint)
- `<video>` (no hint)
- `<object>` (?)
- `<embed>` (?)

**Comment by m2habert** [comment]
There's also a great Vimperator plugin that recognizes text blocks / paragraphs for yanking, leading to a very fast and precise copy & paste experience. I find it to be very useful.Please consider implementing something like it.

**Comment by The-Compiler** [comment]
Another idea from #2078: Something to *download* images

**Comment by The-Compiler** [comment]
Another one from https://github.com/The-Compiler/qutebrowser/issues/2137#issuecomment-263062566: can we have hints for scrollable div's, similar to hinting frames?

**Comment by The-Compiler** [comment]
And one from @haasn: Following the `href` of a link (not clicking), and opening the `src` of an `img` (not clicking it). The latter should probably even be the default behavior of `;i` as that's what dwb is too, and the reason for adding `;i` originally was dwb.

Maybe `F`/`;b` could even always do that instead of doing the current target overriding magic.

**Comment by dkasak** [comment]
Just wanted to mention that "Yank hint anchor URL" is now `;#` in Vimperator, not `;n`.

**Comment by Enzime** [comment]
Another hint mode that might be useful is copying actual image objects to clipboard.

Many HTML5 websites support pasting images now, so this would be a useful feature.

**Comment by The-Compiler** [comment]
And another one from #3298: Rapid hinting for normal mode, which re-enters hinting after loading is finished.

**Comment by jakanaka-evan** [comment]
Just here to say, that some things mentioned here are already possible:
Downloading images: `hint images run download {hint-url}`
Downloading images or links to specific directories: `hint all run download --dest ~/mydir {hint-url}`
Following href without clicking: `hint links run open {hint-url}`
Following image src: `hint images run open {hint-url}` (use -t or -b to open in a new tab or background tab)

Kind of possible:
rapid hinting for yank and appending to clipboard: `hint --rapid all spawn sh -c "echo {hint-url} | xsel -a"` (doesn't work for images for some reason)

**Comment by jakanaka-evan** [comment]
> Note that the image hints currently take the href rather than the src of images IIRC, so that won't work like intended.

If you use `hint links download` or `hint links yank` it uses the href of the surrounding `<a>` tag. But if you use `hint images download/yank` it uses the image's src.

Using the `normal`, `tab-bg` or similar click-based target for images will try to click on the image, effectively doing nothing (or following the `<a>` tag), that's why we use `hint images run open {hint-url}`, since `{hint-url}` seems to successfully get the image `src`, at least in my case.

**Comment by Ram-Z** [comment]
Rapid toggling of checkboxes would be nice. I.e. `hint checkboxes --rapid` or `hint inputs checkboxe --rapid`

**Comment by The-Compiler** [comment]
@Ram-Z A simple `:hint --rapid` already works there (but adds hints for more than checkboxes)

**Comment by Ram-Z** [comment]
Indeed, but I was on a page with a lot of links. `:hint all` was too 
much, pushing the number of chars to 3 and also the hint of the link to 
the right was covering the hint of the checkbox I was interested in.

It's not a big deal... Just thought I'd add it to this list of 
would-be-nice-to-haves.

**Comment by jgkamat** [comment]
Note that #3806 will probably add support for custom selectors for hints!

**Comment by The-Compiler** [comment]
More from #4027: `;k` to kill an element on the page; and a `jseval` hint target would be useful (which sets `this` as the affected element).

**Comment by agguser** [comment]
Please add a hint to reload a broken image.

**Comment by ViktorWalter** [comment]
> Apparently pentadactyl also has a pseudo-frames mode (`F`) which hints scrollable elements like the code block on https://developer.nvidia.com/accelerated-computing-training
> 
> ![bq4](https://cloud.githubusercontent.com/assets/625793/17478207/68cf05e0-5d6a-11e6-9c48-bfbd780e8d87.png)

Is there still not a way to do this in qute? Not even the code box at the top of this thread can be scrolled without mouse as far as I can tell.

**Comment by The-Compiler** [comment]
@ViktorWalter There's #3691 but I haven't had a chance to look at it in detail yet.

**Comment by haasn** [comment]
@Evidlo Can't you already do that with `spawn xdg-open {url}` (or some variation thereof)?

**Comment by Evidlo** [comment]
@haasn No, because `xdg-open` handles all links the same way.  You'd have to download the file to a temporary location and invoke `xdg-open` on that.

**Comment by WloHu** [comment]
Hint mode which opens right click/context menu for elements (VimFx had this).

Is there any way to create custom modes? The list with built-in ones is growing so IMO best idea is to create API for adding custom modes in `config.py` so that people can make and possibly share them.
This way everyone have a choice for picking modes available in their browser. It's also possible for adapting some of them as built-ins.

There are per-url configs and it would be great if the same can be done with custom modes. Combining this with mentioned "open market" for modes it may work in the same way as website skins in userstyles.org .

**Comment by The-Compiler** [comment]
@WloHu This sounds like a job for an extension API and not a config - I opened https://github.com/qutebrowser/qutebrowser-extensions/issues/16.

**Comment by pathetic-lynx** [comment]
What about a hinting mode that spawns the external editor on an input field?
I don't use this feature because the workflow is
1. start normal hinting and choose input field
2. exit input mode
3. press hotkey for ```open-editor```

I've tried to hack this using ```hint inputs hover;; open-editor``` but the element needs to be focused, not hovered.
```hint inputs normal;; open-editor``` does not work either because qb does not wait for the hinting to complete before executing the next command.

**Comment by The-Compiler** [comment]
@pathetic-lynx you can do something like `later 100 open-editor` instead.

**Comment by davidsierradz** [comment]
I want to focus (not hover) some arbitrary element, I'm writing something like:
`config.bind(',,', "hint all run jseval -q let q = document.querySelector('[src*=\"{hint-url}\"],[href*=\"{hint-url}\"]');q.setAttribute('tabIndex', '-1');q.focus()")`, but it's not 100% reliable, any ideas?

**Comment by arza-zara** [comment]
@davidsierradz: Works with https://github.com/qutebrowser/qutebrowser/pull/4715 & `scroll-px`.

**Comment by rbuchberger** [comment]
Adding another requested hint mode: paste from clipboard into selected input. Something like ;p

Thanks!

**Comment by The-Compiler** [comment]
Another one (from IRC): yanking an image's alt text. Perhaps we could even combine that with the `jseval` idea above, so that one could do something like `:link images yank-jseval this.alt` or so.

**Comment by crocket** [comment]
How do I open an image in a new tab or copy an image's URL to clipboard through a hint?

**Comment by The-Compiler** [comment]
Another one [via Reddit](https://www.reddit.com/r/qutebrowser/comments/s60q5p/concatenate_commands_in_keybindings/): Something like `:hint --rapid links spawn`, but with the command being called once, with a list of URLs.

**Comment by The-Compiler** [comment]
Also see #1453 which (despite the title) also talks about adding a hint mode for paragraphs with caret mode.

**Comment by The-Compiler** [comment]
> Hint mode which opens right click/context menu for elements (VimFx had this).

I'm almost 3 years late, but this is actually long possible (though after this was originally posted): 458d280ec64a17946f06050f5f5ad0d5ec506878

**Comment by The-Compiler** [comment]
Another one that came up in IRC, and I think we had an issue about but can't find it: Hints for headings on a page. Then together with #1305, that would allow quickly navigating between headings.

**Comment by Hund** [comment]
Considering the age of this thread, being able to save images via hints, is not something that you're planning to implement support for? :)

**Comment by The-Compiler** [comment]
@Hund An open issue means I intend to do something at some point, otherwise it would be closed. It just so happens that there are over 1000 of those :wink:

**Comment by Hund** [comment]
> @Hund An open issue means I intend to do something at some point, otherwise it would be closed. It just so happens that there are over 1000 of those 😉

Fair point. :D 

I assumed that you would put everything at pause for me, if i asked politely. :P

**Comment by arza-zara** [comment]
@Hund: You can use `:bind ;D hint images spawn wget -P /home/user/Downloads/ {hint-url}`

**Comment by mTvare6** [comment]
There should also be a mode for rapid downloads, currently `;r` breaks when downloading something.

**Comment by The-Compiler** [comment]
@mTvare6 Should work fine if you set `downloads.location.prompt` to false (perhaps temporarily).
